feat: add kid-based signature validation with enhanced error handling

chrisingenhaag opened the pull request:
- add proper kid based single validation for the correct public key instead of trying all available public keys
- if kid is missing, search for unique match of alg for validation
- else fail

chrisingenhaag committed:
feat: add kid-based signature validation with enhanced error handling

chrisingenhaag committed:
feat: add automatic key selection when kid header is absent

Implements fallback key selection based on JWT algorithm and key metadata when kid header is missing. Plugin now attempts to find a single unambiguous matching key by comparing JWT algorithm (RS256/384/512, PS256/384/512, ES256/384/512) with key metadata (alg, kty, use). Rejects tokens when multiple keys match or no matching key is found. Adds key_metadata tracking throughout the key retrieval chain and comprehensive test coverage for various

chrisingenhaag committed:
chore: add explicit docker network configuration for service isolation

Defines kong-net bridge network and assigns all services to it for improved network isolation and DNS resolution. Adds diagnostic network connectivity test script that validates DNS resolution for kong, kc, and httpbin services during test initialization.

chrisingenhaag committed:
chore: add diagnostic logging to GitHub Actions test workflow

Adds Docker version information output and failure diagnostics that display container status and logs for kong, keycloak, and test services when tests fail. Improves troubleshooting capabilities for CI test failures.

chrisingenhaag committed:
fix: add signature validator module to rockspec build configuration

Registers the new validators.signature module in the rockspec build modules list and aligns formatting of validator module entries for consistency.

chrisingenhaag committed:
fix: set all fetching public keys security events to ua221

luispflamminger commented on src/handler.lua:
```diff
@@ -106,19 +107,25 @@ end
+        if kids then
+            key_ids[i] = kids[i]
+        end
     end
```
I don't really understand the reassignment to `key_ids` here. Is this redundant or am I missing something?

chrisingenhaag replied:
It´s part of the return object for later check for available kid´s if the jwt header also contains a kid.

luispflamminger replied:
The redundant reassignment was removed.

luispflamminger commented on src/handler.lua:
```diff
@@ -149,28 +156,39 @@ local function custom_validate_token_signature(conf, jwt, second_call)
     local since_last_update = socket.gettime() - public_keys.updated_at
     if not second_call and since_last_update > conf.iss_key_grace_period then
-        kong.log.debug('Could not validate signature. Keys updated last ' .. since_last_update .. ' seconds ago')
+        kong.log.debug('Could not validate signature with kid-matched key. Keys updated last ' .. since_last_update .. ' seconds ago')
```
```suggestion
        kong.log.debug('Could not validate signature using keys matched by kid or alg. Keys updated last ' .. since_last_update .. ' seconds ago')
```

luispflamminger commented on tests/test_kid_optional.sh:
```diff
@@ -0,0 +1,69 @@
+echo ""
+echo "✅ Kid-optional infrastructure is in place"
+echo "✅ Unit tests verify the kid-optional logic thoroughly"
+echo "✅ Integration tests confirm existing behavior (with kid) still works"
```
I would remove this "instruction style" comment and just put a note/warning that the behavior without kid is currently only tested using unit tests/spec validation.

chrisingenhaag replied:
done

luispflamminger commented on spec/01-unit/validators/signature_spec.lua:
```diff
@@ -0,0 +1,300 @@
+    assert.equals("No matching public key found for token algorithm", err.message)
+  end)
+
+  it("should reject when kids table is missing", function()
```
The current implementation has a fallback to match by alg if kid is missing on the JWT. Should this also work the other way around? So if kid is missing on the entries in the JWKS, it should try and match by algorithm instead of rejecting immediately?

chrisingenhaag replied:
I would assume something is broken if iris creates a jwt with a kid, but then not offering the kids in the jwks certs endpoint. But of course this is an assumption I made to keep the code better maintainable without additional loop and retries for those cases.

luispflamminger commented on src/validators/signature.lua:
```diff
@@ -0,0 +1,137 @@
+      if key_metadata.kty and key_metadata.kty ~= "EC" then
+        return false
+      end
+    end
```
Let's talk about this logic, I don't fully understand why we are not just checking if `jwt.kty == key_metadata.kty`

luispflamminger replied:
We decided to keep the check using a "derived kty" using the alg in the JWT and matching it against the kty in the JWK. This is because alg is optional in the JWK, but kty is not present in the JWT. Matching jwk.alg == jwt.alg is therefore not possible.

luispflamminger commented on spec/01-unit/validators/signature_spec.lua:
```diff
@@ -0,0 +1,300 @@
+    assert.equals("Unable to find public key for token kid", err.message)
+  end)
+
+  it("should reject when kid is not found in public keys", function()
```
```suggestion
  it("should reject when kid is not found in public keys and no match by alg/kty is found", function()
```

chrisingenhaag replied:
done

luispflamminger commented on spec/01-unit/validators/signature_spec.lua:
```diff
@@ -0,0 +1,300 @@
+    assert.equals("No public keys available", err.message)
+  end)
+
+  it("should support PS256 (RSA-PSS) algorithm matching", function()
```
Why are we explicitly checking this algorithm and not others?

chrisingenhaag replied:
yes, this is better and due to a change I made now to not use kty prop for sig alg family matching, this is obsolete.

We´re now only equally checking if a jwt.header.alg is one single matching entry in the keys array[].alg property.

luispflamminger replied:
We decided to keep the check using a "derived kty" using the `alg` in the JWT and matching it against the `kty` in the JWK. This is because `alg` is optional in the JWK, but `kty` is not present in the JWT. Matching `jwk.alg == jwt.alg` is therefore not possible.

luispflamminger reviewed (commented):
I added a couple of comments, mostly for better understanding on my side :)

chrisingenhaag committed:
Update src/handler.lua

Co-authored-by: Luis Pflamminger <luis.pflamminger@gmail.com>

chrisingenhaag committed:
test: remove kty-based key matching and simplify algorithm validation

Removes key type (kty) field from key metadata tracking and eliminates kty-based key matching logic. Simplifies algorithm validation to only check explicit alg field matches when present. Updates test descriptions to reflect that keys match when no alg is specified rather than matching by kty. Removes PS256 algorithm test and kid-optional integration test notes that are now covered by unit tests.

chrisingenhaag committed:
chore: bump plugin version to 1.7.0-1
